support callable in `routes.default_url_options`

ignatiusreza opened the pull request:
to be used when we want to have defaults that are lazily evaluated.

for example, when we want to have `I18n.locale` as part as the path, and we want it to reflect the locale that currently is active when the url is generated

`#default_url_options` in controller solve cases when we're generating the url in controller and views, but it doesn't handle cases when `Rails.application.routes.url_helpers` are included in other classes, which this commit address

rafaelfranca commented:
How that is solved in controllers and views and why we can't use the same implementation for those cases?

ignatiusreza commented:
in master, having something like the following allow us to have dynamic `#default_url_options` in controller and views

```rb
class ApplicationController < ActionController::Base
  def default_url_options
    { locale: I18n.locale }
  end
end
```

since url helpers in controller and view will consider `#default_url_options` defined in controller.. we can also achieve sometime similar in mailers by doing something like:

```rb
class ApplicationMailer < ActionMailer::Base
  def default_url_options
    { locale: I18n.locale }
  end
end
```

since `ActionMailer::Base` inherit from `AbstractController::Base`.. but, it won't work in case we include `Rails.application.routes.url_helpers` in other places, for example in model like:

```rb
class Post < ActiveRecord::Base
  include Rails.application.routes.url_helpers 

  def url(options)
    post_url(self, options)
  end
end
```

since the url_helpers in model are not aware of `#default_url_options` defined in controller..

ignatiusreza commented:
cleaned up the codes and added some more integration tests

ignatiusreza committed:
support callable in `routes.default_url_options`

to be used when we want to have defaults that are lazily evaluated.

for example, when we want to have `I18n.locale` as part as the path,
and we want it to reflect the locale that currently is active when
the url is generated

#default_url_options in controller solve cases when we're generating
the url in controller and views, but it doesn't handle cases when
`Rails.application.routes.url_helpers` are included in other classes,
which this commit address

sgrif commented:
Including URL helpers inside of models is generally considered an anti-pattern, and isn't a great justification for adding a new feature.

rafaelfranca commented:
Yes. I'm with @sgrif on this one. If you need to generate URLs you should do in the context of a controller or view.

ignatiusreza commented:
That's a valid point, and I agree that including url helpers in models is an anti-pattern..

I came up with this feature request, because I had some difficulties not just in model, but also in tests.. I've setup a sample repo that consist of a new rails app with some scaffold sprinkled in with default_url_options in controller and a locale scoped routes.. which hopefully would demonstrate the issue that i'm talking about.. the repo can be found at https://github.com/ignatiusreza/rails-pr-27291/commits/master

This same issue also exists in tests written using rspec or cucumber..
